lib/rules/no-computed-in-data.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * @fileoverview Ensure computed properties are not used in the data()
+ * @author IWANABETHATGUY
+ */
+'use strict'
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+const utils = require('../utils')
+function topOfStack (stack) {
+  console.assert(Array.isArray(stack))
+  return stack[stack.length - 1]
+}
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'disallow computed properties used in the data property',
+      categories: undefined,
+      recommended: false,
+      url: 'https://eslint.vuejs.org/rules/no-computed-in-data.html'
+    },
+    fixable: null, // or "code" or "whitespace"
+    schema: [
+      // fill in your schema
+    ]
+  },
+
+  create: function (context) {
+    let dataAstNode
+    let targetObjectExpression
+    const memberExpressionMap = Object.create(null)
+    const scopeStack = []
+    // if data is a function, should reference its scope, other wise should be undefined
+    let dataPropertyScope
+    return {
+
+      ObjectExpression (node) {
+        if (!dataAstNode) {
+          dataAstNode = node.properties.find(
+            p =>
+              p.type === 'Property' &&
+              p.key.type === 'Identifier' &&
+              p.key.name === 'data'
+          )
+          if (dataAstNode) {
+            targetObjectExpression = node
+            if (dataAstNode.value.type === 'FunctionExpression') {
+              dataPropertyScope = dataAstNode.value
+              scopeStack.push(dataPropertyScope)
+            }
+          }
+        }
+      },
+      ':function' (node) {
+        scopeStack.push(node)
+      },
+      ':function:exit' (node) {
+        scopeStack.pop()
+      },
+      'Property:exit' (node) {
+        if (node === dataAstNode) {
+          dataAstNode = undefined
+        }
+      },
+      MemberExpression (node) {
+        if (dataAstNode && dataPropertyScope === topOfStack(scopeStack)) {
+          // a memberExpression `like this.a.c.d` -> when traverse to c.d we can got the the full name -> this.a.c.d
+          const fullName = utils.parseMemberExpression(node).slice(0, 2).join('.')
+          if (memberExpressionMap[fullName]) {
+            // check if parent node in this array, if true ignore this node, such as `{a: this.a.c.d}`, this traverse function visit order is `this.a.c.d` -> `a.c.d` -> `c.d`
+            const hasParentNodeInArray = memberExpressionMap[fullName].some(nodeInMap => node.parent === nodeInMap)
```

Comment:
I've found an issue with duplicate reports when the chain is deep. Perhaps checking the `parent` is a problem.

```vue
      <script>
      export default {
        data() {
          return {
            a: this.test.foo.bar.baz,
          }
        },
        computed: {
          test() {
            return {}
          },
        }
      }
      </script>
```